# Pillow Palooza





# Data Wrangling and Cleaning

In [1]:
import numpy as np
import pandas as pd

## Importing the Data

In [2]:
# Load "data/airbnb_price.csv" as a DataFrame called prices.

prices = pd.read_csv('data/prices.csv')

# Load "data/airbnb_room_type.xlsx" as a DataFrame called xls, and the first 
# sheet from xls as a DataFrame called room_types.

room_types = pd.read_excel('data/room_types.xlsx')

# Load "data/airbnb_last_review.tsv" as a DataFrame called reviews.

reviews = pd.read_csv('data/reviews.tsv', delimiter='\t')

## Initial Exploration 

In [3]:
prices.head(50)


listing_id        price                          nbhood_full
0         2595  225 dollars                   Manhattan, Midtown
1         3831   89 dollars               Brooklyn, Clinton Hill
2         5099  200 dollars               Manhattan, Murray Hill
3         5178   79 dollars            Manhattan, Hell's Kitchen
4         5238  150 dollars                 Manhattan, Chinatown
5         5295  135 dollars           Manhattan, Upper West Side
6         5441   85 dollars            Manhattan, Hell's Kitchen
7         5803   89 dollars                Brooklyn, South Slope
8         6021   85 dollars           Manhattan, Upper West Side
9         6848  140 dollars               Brooklyn, Williamsburg
10        7097  215 dollars                Brooklyn, Fort Greene
11        7322  140 dollars                   Manhattan, Chelsea
12        7726   99 dollars              Brooklyn, Crown Heights
13        8024  130 dollars                 Brooklyn, Park Slope
14        8025   80 dollars                 Brooklyn, Park Slope
15        8110  110 dollars                 Brooklyn, Park Slope
16        8490  120 dollars         Brooklyn, Bedford-Stuyvesant
17        8505   60 dollars            Brooklyn, Windsor Terrace
18        9518   44 dollars                    Manhattan, Inwood
19        9657  180 dollars              Manhattan, East Village
20        9668   50 dollars                    Manhattan, Harlem
21        9704   52 dollars                    Manhattan, Harlem
22        9782   55 dollars                 Brooklyn, Greenpoint
23        9783   50 dollars                    Manhattan, Harlem
24       10452   70 dollars         Brooklyn, Bedford-Stuyvesant
25       10962   89 dollars                Brooklyn, South Slope
26       11708   85 dollars                   Brooklyn, Bushwick
27       12048   40 dollars           Manhattan, Lower East Side
28       12192   68 dollars              Manhattan, East Village
29       12299  120 dollars                Brooklyn, South Slope
30       12318  135 dollars           Manhattan, Upper West Side
31       12343  150 dollars                    Manhattan, Harlem
32       12627  150 dollars  Brooklyn, Prospect-Lefferts Gardens
33       12937  130 dollars             Queens, Long Island City
34       12940  110 dollars         Brooklyn, Bedford-Stuyvesant
35       13394   80 dollars                Brooklyn, Fort Greene
36       13808   80 dollars         Brooklyn, Bedford-Stuyvesant
37       14287  151 dollars           Manhattan, Upper West Side
38       14290  228 dollars               Brooklyn, Williamsburg
39       14314  144 dollars                 Brooklyn, Greenpoint
40       14322  200 dollars                  Manhattan, Kips Bay
41       14377  150 dollars               Brooklyn, Williamsburg
42       14991  110 dollars           Manhattan, Lower East Side
43       15220   69 dollars            Manhattan, Hell's Kitchen
44       15338   49 dollars                 Brooklyn, Greenpoint
45       15341  180 dollars                      Manhattan, SoHo
46       15385   80 dollars               Brooklyn, Williamsburg
47       15711  250 dollars           Manhattan, Upper East Side
48       16326  200 dollars           Brooklyn, Prospect Heights
49       16421   52 dollars            Manhattan, Hell's Kitchen

In [4]:
print(prices.shape)
print(prices.info())

(25209, 3)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   listing_id   25209 non-null  int64 
 1   price        25209 non-null  object
 2   nbhood_full  25209 non-null  object
dtypes: int64(1), object(2)
memory usage: 591.0+ KB
None


In [5]:
prices[prices.duplicated()==True]

Empty DataFrame
Columns: [listing_id, price, nbhood_full]
Index: []

In [6]:
room_types.head()

listing_id                                description        room_type
0        2595                      Skylit Midtown Castle  Entire home/apt
1        3831            Cozy Entire Floor of Brownstone  Entire home/apt
2        5099  Large Cozy 1 BR Apartment In Midtown East  Entire home/apt
3        5178            Large Furnished Room Near B'way     private room
4        5238         Cute & Cozy Lower East Side 1 bdrm  Entire home/apt

In [7]:
print(room_types.shape)
print(room_types.info())

(25209, 3)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   listing_id   25209 non-null  int64 
 1   description  25199 non-null  object
 2   room_type    25209 non-null  object
dtypes: int64(1), object(2)
memory usage: 591.0+ KB
None


In [8]:
room_types[room_types.duplicated()==True]

Empty DataFrame
Columns: [listing_id, description, room_type]
Index: []

In [9]:
reviews.head(50)

listing_id         host_name      last_review
0         2595          Jennifer      May 21 2019
1         3831       LisaRoxanne     July 05 2019
2         5099             Chris     June 22 2019
3         5178          Shunichi     June 24 2019
4         5238               Ben     June 09 2019
5         5295              Lena     June 22 2019
6         5441              Kate     June 23 2019
7         5803            Laurie     June 24 2019
8         6021           Claudio     July 05 2019
9         6848     Allen & Irina     June 29 2019
10        7097              Jane     June 28 2019
11        7322              Doti     July 01 2019
12        7726  Adam And Charity     June 22 2019
13        8024             Lisel     July 01 2019
14        8025             Lisel  January 01 2019
15        8110             Lisel     July 02 2019
16        8490          Nathalie     June 19 2019
17        8505           Gregory     June 23 2019
18        9518              Shon     June 15 2019
19        9657              Dana    April 19 2019
20        9668   Ssameer Or Trip     June 01 2019
21        9704              Teri     June 14 2019
22        9782            Andrea     June 15 2019
23        9783   Ssameer Or Trip     July 01 2019
24       10452            Angela      May 12 2019
25       10962            Laurie     June 21 2019
26       11708            Tyrome     June 22 2019
27       12048               Ben     July 05 2019
28       12192            Edward     June 21 2019
29       12299             Abdul      May 27 2019
30       12318               Cyn     June 16 2019
31       12343              Earl     June 13 2019
32       12627              Rana     June 05 2019
33       12937           Orestes     July 01 2019
34       12940          Adreinne      May 25 2019
35       13394         Alexander     June 17 2019
36       13808    JT And Tiziana     June 16 2019
37       14287              Joya     July 07 2019
38       14290             James      May 17 2019
39       14314            Jeanne     June 29 2019
40       14322         Francesca    March 25 2019
41       14377            Joanna     June 22 2019
42       14991            Bianca      May 18 2019
43       15220              Luiz     June 09 2019
44       15338            Andrea     June 04 2019
45       15341               Ted     June 29 2019
46       15385          Cristina     June 30 2019
47       15711                 D    March 30 2019
48       16326           Dimitri  January 26 2019
49       16421              Mark      May 16 2019

In [10]:
print(reviews.shape)  
print(reviews.info())

(25209, 3)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   listing_id   25209 non-null  int64 
 1   host_name    25201 non-null  object
 2   last_review  25209 non-null  object
dtypes: int64(1), object(2)
memory usage: 591.0+ KB
None


In [11]:
reviews[reviews.duplicated()==True]

Empty DataFrame
Columns: [listing_id, host_name, last_review]
Index: []

## Initial Observations

- Listing ID should be a string
- incorrect date format in reviews['last_review']
- prices['price'] should be an int, remove 'dollars'
- missing values from reviews['host_name']
- reviews['host_name'] incorrect formatting, contains some 'and', 'or', lack of spaces between name and surname.
- missing values from room_type['description']
- prices['price'] remove outliers
- new column prices['price_per_month']
- prices['nbhood_full'] can be abreviated
- room_type['room_type'] should be standard descriptions. 

## Prices dataframe

### Price column
The price column in the prices DataFrame currently records each value as a string with the currency (dollars) following, i.e. 

We'll use regex to remove the 'dollar' and then convert the datatype of the column to a int.

> price <br>
> 225 dollars <br>
> 89 dollars <br>
> 200 dollars <br>

In [12]:
# Cleaning the Price Column:
# using regex to extract all characters which aren't an int '(\d+)'
# converting the datatype of the column from object to float

prices['price'] = prices['price'].str.extract('(\d+)', 
                                              expand=False).astype(float)

prices.head()

listing_id  price                nbhood_full
0        2595  225.0         Manhattan, Midtown
1        3831   89.0     Brooklyn, Clinton Hill
2        5099  200.0     Manhattan, Murray Hill
3        5178   79.0  Manhattan, Hell's Kitchen
4        5238  150.0       Manhattan, Chinatown

In [13]:
prices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25209 entries, 0 to 25208
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   listing_id   25209 non-null  int64  
 1   price        25209 non-null  float64
 2   nbhood_full  25209 non-null  object 
dtypes: float64(1), int64(1), object(1)
memory usage: 591.0+ KB


### Avg Price column

#### Removing outliers
Three quarters of listings cost $175 per night or less. However, there are some outliers including a maximum price of $7,500 per night. Some listings are even showing as free. We'll remove these from the DataFrame and 
calculate the average price, to the nearest 2 decimal places, storing this as avg_price. Then, calculate the average_price_month. 



In [14]:
# Identify outliers using IQR (Interquartile Range)
Q1 = prices['price'].quantile(0.25)
Q3 = prices['price'].quantile(0.75)
IQR = Q3 - Q1

# Remove outliers in price column
prices = prices[(prices['price'] >= Q1 - 1.5 * IQR) & (prices['price'] <= Q3 + 1.5 * IQR)]

# calculating avg_price
avg_price = prices['price'].mean()

print('$', round(avg_price, 2))

$ 119.15


### Avg Price per Month column

We know how much a listing costs, on average, per night, but it would be useful to have a benchmark for comparison. According to [Zumper](https://www.zumper.com/rent-research), a 1 bedroom apartment in New York City costs, on average, $3,100 per month. 

So we'll convert the per-night prices of our listings into monthly costs, so we can compare to the private market.



In [15]:
# new column: monthly_price
prices['monthly_price'] = prices['price']*30

# calculate avergae monthly price 
avg_monthly_price = prices['monthly_price'].mean()

print('$', round(avg_monthly_price, 2))

$ 3574.5


## Room types dataframe

### Room Type column

The room_type column has several variations for private room listings. 
We'll dive deeper into this column to clean the breakdown of listings by room type. 

In [16]:
# Using regex to and replacing several variations of room type with a standard description. 

room_types['room_type'] = room_types['room_type'].str.lower()

room_types['room_type'].value_counts()

entire home/apt    13266
private room       11356
shared room          587
Name: room_type, dtype: int64

## Reviews dataframe

### Last review column
The last_review column contains the date of the last review in the format of "Month Day Year" e.g., May 21 2019. 
We'll change this format to a pandas datetime.


In [17]:
# incorrect date format in reviews['last_review']

reviews['last_review'] = pd.to_datetime(reviews['last_review'])
reviews.head()

listing_id    host_name last_review
0        2595     Jennifer  2019-05-21
1        3831  LisaRoxanne  2019-07-05
2        5099        Chris  2019-06-22
3        5178     Shunichi  2019-06-24
4        5238          Ben  2019-06-09

## Host Name Column

Clean this column to standardize name format. 

In [27]:
# format host name

# Add a space between a lowercase letter and an uppercase letter
reviews['host_name'] = reviews['host_name'].str.replace(r'([a-z])([A-Z])', r'\1 \2')

# Capitalize the first and second words
reviews['host_name'] = reviews['host_name'].str.title()

# Replace '&', 'AND', 'and', 'Or', 'OR', 'And', 'or' with '&'
reviews['host_name'] = reviews['host_name'].str.replace(r'(&| AND | and | Or | OR | And | or )', ' & ')

/var/folders/1m/hvnyjbcx2rxfb9d__rb_m7h00000gp/T/ipykernel_76375/1448132106.py:4: FutureWarning: The default value of regex will change from True to False in a future version.
  reviews['host_name'] = reviews['host_name'].str.replace(r'([a-z])([A-Z])', r'\1 \2')
/var/folders/1m/hvnyjbcx2rxfb9d__rb_m7h00000gp/T/ipykernel_76375/1448132106.py:10: FutureWarning: The default value of regex will change from True to False in a future version.
  reviews['host_name'] = reviews['host_name'].str.replace(r'(&| AND | and | Or | OR | And | or )', ' & ')


listing_id                      host_name last_review
0         2595                       Jennifer  2019-05-21
1         3831                   Lisa Roxanne  2019-07-05
2         5099                          Chris  2019-06-22
3         5178                       Shunichi  2019-06-24
4         5238                            Ben  2019-06-09
5         5295                           Lena  2019-06-22
6         5441                           Kate  2019-06-23
7         5803                         Laurie  2019-06-24
8         6021                        Claudio  2019-07-05
9         6848  Allen         &         Irina  2019-06-29
10        7097                           Jane  2019-06-28
11        7322                           Doti  2019-07-01
12        7726   Adam        &        Charity  2019-06-22
13        8024                          Lisel  2019-07-01
14        8025                          Lisel  2019-01-01
15        8110                          Lisel  2019-07-02
16        8490                       Nathalie  2019-06-19
17        8505                        Gregory  2019-06-23
18        9518                           Shon  2019-06-15
19        9657                           Dana  2019-04-19
20        9668   Ssameer        &        Trip  2019-06-01
21        9704                           Teri  2019-06-14
22        9782                         Andrea  2019-06-15
23        9783   Ssameer        &        Trip  2019-07-01
24       10452                         Angela  2019-05-12
25       10962                         Laurie  2019-06-21
26       11708                         Tyrome  2019-06-22
27       12048                            Ben  2019-07-05
28       12192                         Edward  2019-06-21
29       12299                          Abdul  2019-05-27
30       12318                            Cyn  2019-06-16
31       12343                           Earl  2019-06-13
32       12627                           Rana  2019-06-05
33       12937                        Orestes  2019-07-01
34       12940                       Adreinne  2019-05-25
35       13394                      Alexander  2019-06-17
36       13808     Jt        &        Tiziana  2019-06-16
37       14287                           Joya  2019-07-07
38       14290                          James  2019-05-17
39       14314                         Jeanne  2019-06-29
40       14322                      Francesca  2019-03-25
41       14377                         Joanna  2019-06-22
42       14991                         Bianca  2019-05-18
43       15220                           Luiz  2019-06-09
44       15338                         Andrea  2019-06-04
45       15341                            Ted  2019-06-29
46       15385                       Cristina  2019-06-30
47       15711                              D  2019-03-30
48       16326                        Dimitri  2019-01-26
49       16421                           Mark  2019-05-16

## Joining the dataframes

* Merge the three DataFrames to make any future analysis easier to conduct. 

* Remove any observations with missing values and check for duplicates.

* Convert listing ID to string



In [29]:
df = pd.merge(reviews, prices, on='listing_id')
df = pd.merge(df, room_types, on='listing_id')

df.head()

listing_id     host_name last_review  price                nbhood_full  \
0        2595      Jennifer  2019-05-21  225.0         Manhattan, Midtown   
1        3831  Lisa Roxanne  2019-07-05   89.0     Brooklyn, Clinton Hill   
2        5099         Chris  2019-06-22  200.0     Manhattan, Murray Hill   
3        5178      Shunichi  2019-06-24   79.0  Manhattan, Hell's Kitchen   
4        5238           Ben  2019-06-09  150.0       Manhattan, Chinatown   

   monthly_price                                description        room_type  
0         6750.0                      Skylit Midtown Castle  entire home/apt  
1         2670.0            Cozy Entire Floor of Brownstone  entire home/apt  
2         6000.0  Large Cozy 1 BR Apartment In Midtown East  entire home/apt  
3         2370.0            Large Furnished Room Near B'way     private room  
4         4500.0         Cute & Cozy Lower East Side 1 bdrm  entire home/apt

In [42]:
df['listing_id'] = df['listing_id'].astype('string')

df['description'] = df['description'].fillna('No description')
df['host_name'] = df['host_name'].fillna('Unknown.')

df.isnull().sum()

listing_id       0
host_name        0
last_review      0
price            0
nbhood_full      0
monthly_price    0
description      0
room_type        0
dtype: int64

## Analyzing listing prices by NYC borough

* Initial analysis to understand the difference in listing prices between the [New York City boroughs](https://en.wikipedia.org/wiki/Boroughs_of_New_York_City).
* Extract the information from nbhood_full  and store in a new column, borough, for analysis.

In [47]:
df['borough'] = df['nbhood_full'].str.split(',').str.get(0)

df.head()

listing_id     host_name last_review  price                nbhood_full  \
0       2595      Jennifer  2019-05-21  225.0         Manhattan, Midtown   
1       3831  Lisa Roxanne  2019-07-05   89.0     Brooklyn, Clinton Hill   
2       5099         Chris  2019-06-22  200.0     Manhattan, Murray Hill   
3       5178      Shunichi  2019-06-24   79.0  Manhattan, Hell's Kitchen   
4       5238           Ben  2019-06-09  150.0       Manhattan, Chinatown   

   monthly_price                                description        room_type  \
0         6750.0                      Skylit Midtown Castle  entire home/apt   
1         2670.0            Cozy Entire Floor of Brownstone  entire home/apt   
2         6000.0  Large Cozy 1 BR Apartment In Midtown East  entire home/apt   
3         2370.0            Large Furnished Room Near B'way     private room   
4         4500.0         Cute & Cozy Lower East Side 1 bdrm  entire home/apt   

     borough  
0  Manhattan  
1   Brooklyn  
2  Manhattan  
3  Manhattan  
4  Manhattan

## Price range by borough

* categorize listings based on whether they fall into specific price ranges, and view this by borough.

* percentiles and labels to create a new column, price_range, in the DataFrame. 



In [56]:

price_ranges = {
    (0, 69): '$0-69',
    (70, 175): '$70-175',
    (176, 350): '$176-350',
    (351, float('inf')): '>$350'
}

def categorize_price(price):
    for price_range, label in price_ranges.items():
        if price >= price_range[0] and price <= price_range[1]:
            return label

df['price_range'] = df['price'].map(categorize_price)
df['labels'] = df['price_range'].replace(price_ranges.values(), ['Budget', 'Average', 'Expensive', 'Extravagant'])



df.head()

listing_id     host_name last_review  price                nbhood_full  \
0       2595      Jennifer  2019-05-21  225.0         Manhattan, Midtown   
1       3831  Lisa Roxanne  2019-07-05   89.0     Brooklyn, Clinton Hill   
2       5099         Chris  2019-06-22  200.0     Manhattan, Murray Hill   
3       5178      Shunichi  2019-06-24   79.0  Manhattan, Hell's Kitchen   
4       5238           Ben  2019-06-09  150.0       Manhattan, Chinatown   

   monthly_price                                description        room_type  \
0         6750.0                      Skylit Midtown Castle  entire home/apt   
1         2670.0            Cozy Entire Floor of Brownstone  entire home/apt   
2         6000.0  Large Cozy 1 BR Apartment In Midtown East  entire home/apt   
3         2370.0            Large Furnished Room Near B'way     private room   
4         4500.0         Cute & Cozy Lower East Side 1 bdrm  entire home/apt   

     borough price_range     labels  
0  Manhattan    $176-350  Expensive  
1   Brooklyn     $70-175    Average  
2  Manhattan    $176-350  Expensive  
3  Manhattan     $70-175    Average  
4  Manhattan     $70-175    Average


* group the data and count frequencies for listings in each price range by borough.

| label      | price     | 
|------------|-----------|
| budget     | \$0-69    | 
| average    | \$70-175  |
| expensive  | \$176-350 | 
| extravagant| >\$350    |

In [54]:
# Group the data by 'borough' and 'price_range', and count the frequencies
grouped_df = df.groupby(['borough', 'price_range', 'labels']).size().reset_index(name='frequency')

grouped_df.head(50)


borough price_range     labels  frequency
0           Bronx       $0-69     Budget        382
1           Bronx    $176-350  Expensive         24
2           Bronx     $70-175    Average        286
3        Brooklyn       $0-69     Budget       3207
4        Brooklyn    $176-350  Expensive       1399
5        Brooklyn     $70-175    Average       5534
6       Manhattan       $0-69     Budget       1150
7       Manhattan    $176-350  Expensive       2926
8       Manhattan     $70-175    Average       5289
9          Queens       $0-69     Budget       1632
10         Queens    $176-350  Expensive        282
11         Queens     $70-175    Average       1505
12  Staten Island       $0-69     Budget        124
13  Staten Island    $176-350  Expensive         20
14  Staten Island     $70-175    Average        123

## Exporting the clean datasets